docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,206 @@
+
+__New Seed__ is the seed to which the control plane is being migrated
+
+
+## Resources and controller state which have to be migrated between two seeds:
+
+### Secrets
+
+Gardener has preconfigured lists of needed secrets which are generated when a shoot is created and deployed in the seed. Following is a minimum set of secrets which must be migrated to the New Seed. Other secrets can be regenerated from them.
+
+- ca
+- ca-front-proxy
+- static-token
+- ca-kubelet
+- ca-metrics-server
+- etcd-encryption-secret
+- kube-aggregator
+- kube-apiserver-basic-auth
+- kube-apiserver
+- service-account-key
+- ssh-keypair
+
+### Custom Resources and state of extension controllers
+
+Gardener deploys custom resources to the Host Seed cluster which are reconciled by extension controllers. The state of these controllers and any additional resources they create is independent of gardener and must also be migrated to the New Seed. Following is a list of custom resources, and the state which is generated by them that has to be migrated.
+
+- BackupBucket: nothing relevant for migration
+- BackupEntry: nothing relevant for migration
+- ControlPlane: nothing relevant for migration
+- DNSProvider/DNSEntry: nothing relevant for migration
+- Extensions: migration of state needs to be handled individually
+- Infrastructure: terraform state
+- Network: nothing relevant for migration
+- OperatingSystemConfig: nothing relevant for migration
+- Worker: Machine-Controller-Manager related objects: machineclasses, machinedeployments, machinesets, machines
+
+This list depends on the currently installed extensions and can change in the future
+
+## Proposal
+
+### CRD on the garden cluster
+
+The garden cluster has a new CustomResource. It contains all the required data described above so that the the control plane can be recreated on the New Seed.
+The metadata is separated into two sections. The first is generated by the Gardener Controller Manager and then either used to generate new resources (e.g secrets) or data in it is directly deployed to the Shoot's control plane on the New Seed.
+
+The second is generated by the extension controllers in the seed and then synced to the shoot by a new ShootState controller (described below).
+
+```
+apiVersion: core.gardener.cloud/v1alpha1
+kind: ShootState
+metadata:
+  name: my-shoot
+  namespace: garden-core
+gardener:
+- name: ca
+  data:
+    ca.crt: ...
+    ca.key: ...
+- name: ssh-keypair
+  data:
+    id_rsa: ...
+- name:
+...
+extensions:
+- kind: Infrastructure
+  state: ... (Terraform state)
+- kind: ControlPlane
+  purpose: normal
+  state: ... (Certificates generated by the extension)
+- kind: Worker
+  state: ... (Machine objects)
+```
+
+### Gardener extensions changes
+
+All extension controllers which support state migration must have a new `status.state` field and support an annotation `core.gardener.cloud/operation=restore` in the respective Custom Resources.
```

Comment:
That reads as if not every extension controller has to support migration, but if that's not the case, then how would Gardener know whether migration is feasible and do we even want to allow controllers that don't support migration and why? It should be a general rule, that all have to comply or else they don't pass the certification.

Reply:
Text should be
```
All extension controllers which require state migration ...
```

As an example for a controller which does not require state migration: 
the current backupentry controller does not require to save its state so it will just have `status.state=nil` and if the BackupEntry CR is annotated with `core.gardener.cloud/operation=restore` normal reconciliation should be triggered. In future implementations of that controller its state might have to be saved and the reconciliation/migration flow in the gardenlet/GCM should not have to change.

Reply:
👍